### ISSUE TITLE
HTTP/3 support

### ISSUE THREAD

**Issue by gdubicki** [haproxy/haproxy#680]
## Output of `haproxy -vv` and `uname -a`

Future version

## What should haproxy do differently? Which functionality do you think we should add?

Add [HTTP/3](https://en.wikipedia.org/wiki/HTTP/3) support.

(keywords: HTTP3, QUIC)

**Comment by wtarreau** [comment]
That's still a work in progress. The QUIC part is the one requiring the most work.

**Comment by wtarreau** [comment]
Also, for posterity, please note that the HTTP/3 & QUIC standards are not defined by Wikipedia but by the IETF QUIC working group: [https://tools.ietf.org/wg/quic/](https://tools.ietf.org/wg/quic/).

**Comment by wtarreau** [comment]
For those interested, Fred has recently uploaded his QUIC development branch here: https://github.com/haproxytech/quic-dev

Note that this is still early developer's work and isn't usable for anything but helping to develop the protocol.

**Comment by ageric** [comment]
There's a good implementation of the QUIC protocol at https://github.com/h2o/quicly in case you run into issues.

**Comment by wtarreau** [comment]
Yep, thanks. Each stack will be very specific to its environment and can not necessarily be easy to port to another one. However the availability of other stacks is critical for interoperability tests and to clear out some doubts about certain specs interpretations.

**Comment by myjimmy** [comment]
I found the fact that HAProxy supports HTTP/3 (QUIC) through **[this link](https://www.haproxy.com/blog/announcing-haproxy-2-3/#connection-improvements)**.
But I can't find any data related to this anywhere.

**I'd like to know that HAProxy is ready to support HTTP/3 (QUIC) now.**

**Comment by TimWolla** [comment]
Note: The above message is a duplicate of this mailing list thread: https://www.mail-archive.com/haproxy@formilux.org/msg39558.html.

**Comment by TimWolla** [comment]
@myjimmy I'm not seeing where you read that in the link you gave. The link says that some of the *preparations* to support QUIC have been concluded. HAProxy 2.3 does not have support for QUIC.

**Comment by wtarreau** [comment]
Confirmed, no QUIC yet but developments are still ongoing. Maybe would have something experimental by 2.4 but it's far from being sure.

**Comment by myjimmy** [comment]
@TimWolla, @wtarreau, Thanks for your reply.
It seems that I'm thinking something wrong about supporting HTTP/3 (QUIC) in HAProxy.
When will this feature of HAProxy release?

**Comment by wtarreau** [comment]
As I explained, it's still not supported and we might possibly have it in experimental state around the 2.4 release. I'm more confident in 2.5 though. In any case, the first version will be of the same level of advancement as H2 was in 1.8.

You cannot yet test it with USE_QUIC=1 because this was essentially added for QUIC developers to have some part of their code already merged in instead of constantly seeing it broken by API changes from others.

**Comment by gdubicki** [comment]
QUIC is described as [RFC 9000](https://datatracker.ietf.org/doc/html/rfc9000) now.

**Comment by git001** [comment]
HTTP/3 is not yet finished according to bagder https://daniel.haxx.se/blog/2021/05/27/quic-is-rfc-9000/

**Comment by wtarreau** [comment]
That's a detail, in practice it's ready, it's just a matter of placing the specs in the right order to avoid confusion later.

**Comment by vigorousdeveloper** [comment]
Hi, guys,
Now the haproxy-2.4-dev5 supports the HTTP/3 or else?

Thank you in advance.

**Comment by wtarreau** [comment]
I guess you meant "2.5-dev5" instead since 2.4 was already released a while ago. But no, not yet, it's still a very difficult work in progress but I'm told that we're getting closer to seeing our first very simple requests.

**Comment by wtarreau** [comment]
For those interested, the last series of patches that were merged and that will be in 2.5-dev8 allows to pass some requests to the servers, but be careful, that's only for developers, resource management is still far from being sufficient, and important bugs and limitations are still there. It just shows that significant progress was done, it already allows to perform the test documented here: https://twitter.com/WillyTarreau/status/1441051288997924870 . Please just do not deploy in prod for now!

**Comment by OrkhanAlikhanov** [comment]
Hello @wtarreau, do you have any updates to share on the status of this incredible work? 🙂

**Comment by wtarreau** [comment]
Well, these updates are shared weekly :-)  Here's the latest one: https://www.mail-archive.com/search?l=haproxy%40formilux.org&q=announce+subject%3A%22[ANNOUNCE]+haproxy-2.6-dev8%22+-subject%3A%22re:%22
In short, https://haproxy.org is already being served in HTTP/3 over QUIC using the very latest 2.6-dev and has been running fine for a few weeks now. Some limitations and bugs are of course being addressed daily but it has reached a state where we consider that the risks are low enough for a monitored deployment.

We expect that 2.6 will have it "ready for production run by experts, but with experimental status", which means that it will be fine for production run by admins who understand what they're doing, and we'll do our best to fix issues there, but that due to it being very young we cannot rule out the risk that some issues will require that it has to be disabled for the time needed to fix it. I.e. if you want to try to use it for your site or hosted sites, that will be fine, but do not use that as an important differentiator your business depends on because you may have to disable it a few times during its life.

Stay tuned ;-)

**Comment by wtarreau** [comment]
Closing since implemented in stable 2.6.

**Comment by iuridiniz** [comment]
I'm trying to maintain docker images with haproxy+QUIC (http/3) under  https://hub.docker.com/r/iuridiniz/haproxy

**Comment by wtarreau** [comment]
Nice, thank you! FYI recently a new "quic_enabled" sample fetch function was added so that you can condition your alt-svc rule with it. This will allow you to globally disable QUIC from the global section if desired (in case you suspect a problem for example) and automatically turn off the alt-svc rules without having to comment them out. It also works when quic is not compiled it (it returns false). You may want to have a look at the doc for more details.

Also for those interested, there are also some native packages with QUIC for RHEL/CentOS here:  https://github.com/haproxy/wiki/wiki/Packages

**Comment by mr-manuel** [comment]
Is QUIC still in experimental phase or will it never be included in the main native package like the Debian package?

**Comment by lukastribus** [comment]
The problem is not haproxy but the ssl library. OpenSSL has decided to not use existing QUIC implementations and do everything from scratch, which means that it will be some time before a) OpenSSL releases a new major version with QUIC support, b) haproxy adapts the new API and c) Debian actually ships those releases (both OpenSSL and Haproxy).

Until then, vendoring in a external openssl fork with quic is required.

**Comment by kiler129** [comment]
cURL author has s good summary of the whole OpenSSL & QUIC saga: https://daniel.haxx.se/blog/2021/10/25/the-quic-api-openssl-will-not-provide/

In short, it looks like it will take a few years for the upstream OpenSSL to implement QUIC. In addition, the goal is to have a different API than one in BoringSSL/QUICSSL fork... so as @lukastribus said the support is experimental as it uses a form.

The question is how experimental is the experimental in haproxy specifically? A good chunk of the world is using QUIC with BoringSSL/QUICSSL fork.

**Comment by wtarreau** [comment]
In fact as Lukas explained, the situation is complicated. For us, QUIC is stable in that it's actively maintained and usable in production, but we know that being a very young protocol for which attacks have not yet developed, and implemented on a very young code base, there are still risks that at any moment we have to say "disable it for now, we're trying to figure a fix". When 2.7 was released we considered that the lack of support for the reload could be a bit limitating for some deployments, and the usage in field was still less than a year so we preferred to say "use it but don't base your business on it". Now things have significantly improved and we intend to reach a state of a completely mainstream implementation for 2.8. Does it mean that there will likely be more bugs there than in H2 ? Sure. Just like there are more in H2 than H1. That won't make it not prod-ready regardless.

And as Lukas explained, everyone is still suffering from the big mess and huge confusion that the openssl team has created by letting a critical patchset rot for 3+ years then rejecting it in committee despite everyone else adopted it. And there is no way I can forsee anything usable coming from them in the next decade for the sole reason that they're aiming waaaaay beyond their capabilities as they prove version after version by further degrading their product despite accumulating delays. They've already ruined 3.0's performance compared to 1.1 and are trying hard to recover what can be still be saved but it's obvious to me they're currently exploring areas they have zero knowledge about, and this mixed with transport will be yet another dimension. It would be foolish to just imagine that anything coming from there could be usable for more than just displaying "hello world" without revisiting the whole project's governance and organization; they're [proudly inflicting to themselves](https://www.openssl.org/policies/omc-bylaws.html#OTC) what is considered in the rest of the industry as [something pejorative: "design by committee"](https://en.wikipedia.org/wiki/Design_by_committee). My guess is that the team is just no more interested at all in any technical work but only in paperwork, compliance and such stuff, and that FIPS was a great opportunity for them to have "fun" making 3.0. And showing so little love for what made their project succeed, and [so much contempt for their users](https://github.com/openssl/openssl/pull/8797#issuecomment-942323152) just makes me consider that this project's long and painful death is already engaged and that this time it's irreversible, not like the few previous times, because they've accumulated too many issues to be addressed in parallel to survive.

What this means for now is that the solution **will** come from alternatives. Either an officially maintained fork (QuicTLS is maintained by brave volunteers, how long will they be able to do that, especially with no distro willing to package it, I don't know). Others are not much present (BoringSSL is just rolling releases for Chrome, Libressl lives 1 year max and seems to be disappearing from distros, GnuTLS has a totally different API, WolfSSL currently seems to be the most likely candidate to replace OpenSSL everywhere). At least we need to convince distros that it's also in their interest to enable QUIC support (either by merging the QuicTLS patches in their OpenSSL packages or by providing a TLS stack that is maintained by people living in the 21th century). But replacing it will definitely take time. For sure, OpenSSL's team is helping that a lot by destroying their own product like they're currently doing, but that should not be the right way to proceed :-(  If at least some of the foundation's sponsors would start to question what's being done with their money, it could possibly speed the process up a little bit...

**Comment by lukastribus** [comment]
What's your opinion about rustls or rust dependencies in general?

Curl does have a rust backend now, while it's author actually works for wolfssl:
https://daniel.haxx.se/blog/2022/02/01/curl-with-rust/

So I wonder how far fetched this would be for haproxy.

**Comment by wtarreau** [comment]
I don't have a particular opinion and I know that William was interested in having a look. The thing is that our features directly derive from the openssl-inherited API and I suspect it will always be difficult to reimplement something using a different API. It turns out that wolfssl seems to have good openssl API compatibility, but it's difficult to figure what's complete and what is not. We've reached the point where it works with both tcp and QUIC but some regtests had to be disabled as they were failing. Also we noticed small differences in behavior that we don't know whether they're a bug on either side or just different interpretations of a gray area. The problem I'm foreseeing is users getting confused by possibly different error codes that we interpret differently, or event counters that do not translate what happened. But at least it's extremely fast, **does** scale with threads (not like OpenSSL whose performance is literally divided by the number of threads) and their team seems responsive. Overall I would like to help them become more mainstream. If we could make it possible for next year's LTS distros to ship with it in a form that allows haproxy to build with it by default, it would be awesome, and we could definitely say goodbye to the defunct OpenSSL. And maybe it's not that far-fetched.

**Comment by vincentbernat** [comment]
As for Debian, WolfSSL is available (currently, latest version would be in bookworm).

**Comment by wtarreau** [comment]
Yes, but there's some work to do on their side as well to make it easier to use. It comes from the embedded world and there are tens (maybe even one hundred) of optional features to turn on or off at build time. As much as I would like to see haproxy, ngtcp2, curl, Apache, maybe NGINX etc being supported by the same build of the same lib, I'm not sure that it currently is possible at all. There are also some build options that set the cache size, or the type of CPU to enable or disable crypto acceleration, etc. Some of them need to be made more compatible and others should at least have some boot-time detection with a safe generic fallback. If collectively we can provide them such a feedback about the minimum requirements to support a few packaged programs that need high performance and QUIC, it could possibly help them adjust the product a little bit to improve integration. At the moment it remains non-trivial to configure. On the other hand the project is super small, it builds in 5 seconds or so, so one possibility (that I'd rather avoid) could also be to embed a copy of it in relevant software. But I'm not seeing us switching in a default build until all features are entirely supported anyway.

**Comment by njsmith** [comment]
BoringSSL is the one I usually see people talking about as the reliable OpenSSL fork -- I'm curious if you could elaborate on what you mean by it being "just rolling releases for Chrome"? The lack of maintenance guarantees is certainly a major downside, but Google's replaced OpenSSL with it not just in Chrome but in handsets, in their datacenters, etc., so the drop-in compatibility is demonstrably pretty solid in practice. And it's diverged from upstream OpenSSL in lots of ways, all of them AFAICT for the better.

**Comment by wtarreau** [comment]
The problem is clearly that you can't get a stable version, and there's even no guarantee of a minimally stable API. Once you're forced to upgrade, it often breaks the build and you're stuck with no solution. This is the main problem. We had to disable it from the CI because of this :-(

**Comment by richsalz** [comment]
QuicTLS (https://github.com/quictls/openssl/) will be stable for a very long time. It is the base for what Akamai uses internally.

HOWEVER, we are upset with the direction the OpenSSL project is taking. Willy explained some of the reasons above, https://github.com/haproxy/haproxy/issues/680#issuecomment-1433118828, but it goes deeper and is more fundamental.

Akamai is interested in creating an Apache project from a fork. If you are interested or want to learn more (I have a very rough draft Apache Incubator proposal), email me. rsalz at akamai.com

**Comment by wtarreau** [comment]
@richsalz do you think there will be plans to roll back the parts that ruined the performance between 1.1 and 3.0 ? On some use cases the performance was divided by 200...  We're in a situation where we'll have to maintain our own 1.1.1 fork for some users just because 3.0 isn't usable at all.

**Comment by richsalz** [comment]
We want to address the issue, yes. Of course, it will depend on what the project community wants to do.  (We can't use 1.1 because only 3.0 and later have the Apache license.)  But please spread the word to anyone who might be interested in apache-tls.

**Comment by wtarreau** [comment]
Yes I figured the licensing problem as well and can understand. We (at haproxy) were interested in providing some help to improve openssl's performance by removing most of the unneeded locking. That was before the project took the totally absurd directions that we all know. Well, if finally realistic support of features users need becomes an option again, we'll be pleased to be able to provide some help as we already do for other side projects (like the QAT engine or wolfSSL for example). That's a great news, Rich, thanks for all that you and Nick are doing, and for keeping us informed of a possible brighter future!

**Comment by paulocoghi** [comment]
@richsalz I believe `apache-tls` is the best path to follow, despite the (likely) initial pain.

It will give oxygen to many important contributors (as seen in https://github.com/openssl/openssl/pull/8797) and, most probably, will become the de-facto SSL library.

**Comment by paulocoghi** [comment]
I see this as a partially similar replay of what happened with MySQL and its MariaDB fork. It wasn't easy, but it was necessary and highly beneficial.

Major contributors migrated to MariaDB, and soon after it became the central hub where improvements happen and come first.

**Comment by git001** [comment]
Cite from https://github.com/haproxy/haproxy/issues/680#issuecomment-1434776627  
> We want to address the issue, yes. Of course, it will depend on what the project community wants to do. (We can't use 1.1 because only 3.0 and later have the Apache license.) But please spread the word to anyone who might be interested in apache-tls.

@richsalz as I haven't seen the `apache-tls` for now in the project list https://apache.org/index.html#projects-list please can you share the project link after it's available.

**Comment by richsalz** [comment]
We will be going through the Incubator, and right now we're looking for partners to join.  Ping me rsalz at akamai if you are interested and/or curious.
